services/vision/obstaclesdepth/obstacles_depth.go

```diff
@@ -209,24 +211,39 @@ func (o *obsDepth) obsDepthWithIntrinsics(ctx context.Context, src camera.VideoS
 	w, h := dm.Width(), dm.Height()
 	o.dm = dm
 
-	obstaclePoints := make([]image.Point, 0, w*h/sampleN)
+	var wg sync.WaitGroup
+	obstaclePointChan := make(chan image.Point)
+
 	for i := 0; i < w; i += sampleN {
-		for j := 0; j < h; j++ {
-			candidate := image.Pt(i, j)
-		obs: // for every sub-sampled point, figure out if it is an obstacle
-			for l := 0; l < w; l += sampleN { // continue with the sub-sampling
-				for m := 0; m < h; m++ {
-					compareTo := image.Pt(l, m)
-					if candidate == compareTo {
-						continue
-					}
-					if o.isCompatible(candidate, compareTo) {
-						obstaclePoints = append(obstaclePoints, candidate)
-						break obs
+		wg.Add(1)
+		go func(i int) {
+			defer wg.Done()
+			for j := 0; j < h; j++ {
+				candidate := image.Pt(i, j)
+			obs: // for every sub-sampled point, figure out if it is an obstacle
+				for l := 0; l < w; l += sampleN { // continue with the sub-sampling
+					for m := 0; m < h; m++ {
+						compareTo := image.Pt(l, m)
+						if candidate == compareTo {
+							continue
+						}
+						if o.isCompatible(candidate, compareTo) {
+							obstaclePointChan <- candidate
+							break obs
+						}
 					}
 				}
 			}
-		}
+		}(i)
+	}
+	goutils.PanicCapturingGo(func() {
+		wg.Wait()
+		close(obstaclePointChan)
+	})
```

Comment:
why wait in a routine? is this likely to panic? I'd prefer the loop below this to be in a routine more than the wait/close, to make sure that we close/finish waiting before we leave this function.

alternatively you can maybe do something like [here](https://stackoverflow.com/a/54534707), though we don't know how many candidates we'll be putting into the channel

Reply:
This isn't likely to panic (but I thought this was better than a raw `go func() ` (maybe I should use the ManagedGo here?) This is the closer routine to make sure we finish up.